File: app/src/main/java/com/example/myappcontacts/presentation/contactslist/adapters/ContactsListRecyclerAdapter.java
package com.example.myappcontacts.presentation.contactslist.adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.myappcontacts.R;
import com.example.myappcontacts.data.dao.contacts.db.ContactsModel;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ContactsListRecyclerAdapter extends RecyclerView.Adapter<ContactsListRecyclerAdapter.ContactsHolder> implements ItemTouchHelperAdapter {
    private List<ContactsModel> mContactsList;
    private OnItemListener mOnItemListener;

    public ContactsListRecyclerAdapter(OnItemListener onItemListener) {
        mContactsList = new ArrayList<>();
        mOnItemListener = onItemListener;
    }

    @NonNull
    @Override
    public ContactsHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.item_contacts_list, viewGroup, false);
        return new ContactsHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ContactsHolder contactsHolder, int i) {
        contactsHolder.bindContact(mContactsList.get(i));
    }

    @Override
    public int getItemCount() {
        return mContactsList.size();
    }

    public void updateList(List<ContactsModel> contactsList) {
        mContactsList.clear();
        mContactsList.addAll(contactsList);
        notifyDataSetChanged();
        Log.i("MY_TAG", "notifyDataSetChanged() адаптера сработал");
    }

    @Override
    public void onItemDismiss(int position) {
        mOnItemListener.onItemSwipe(mContactsList.get(position).getContactId());
        mContactsList.remove(position);
        notifyDataSetChanged();
    }

    class ContactsHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private UUID mContactId;

        @BindView(R.id.contact_name)
        TextView mContactNameTextView;
        @BindView(R.id.contact_have_telephone)
        ImageView mContactTelephone;

        public ContactsHolder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
        }

        void bindContact(ContactsModel contactsModel) {
            mContactId = contactsModel.getContactId();
            String firstName = contactsModel.getFirstName();
            String lastName = contactsModel.getLastName();
            if (firstName.equals("") && lastName.equals("")) {
                mContactNameTextView.setText(R.string.no_name);
            } else {
                String fullName = lastName + " " + firstName;
                mContactNameTextView.setText(fullName.trim());
            }
            if (contactsModel.getTelNumber().equals("")) {
                mContactTelephone.setVisibility(ImageView.GONE);
            } else {
                mContactTelephone.setVisibility(ImageView.VISIBLE);
            }
        }

        @Override
        public void onClick(View v) {
            mOnItemListener.onItemClick(mContactId);
        }
    }

    public interface OnItemListener {
        void onItemClick(UUID contactId);

        void onItemSwipe(UUID contactId);
    }
}

File: app/src/main/java/com/example/myappcontacts/presentation/contactslist/adapters/ContactsListRecyclerAdapter.kt
package com.example.myappcontacts.presentation.contactslist.adapters

import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import com.example.myappcontacts.R
import com.example.myappcontacts.data.dao.contacts.db.ContactsModel
import kotlinx.android.synthetic.main.item_contacts_list.view.*
import java.util.*

internal const val TAG = "TEST"

class ContactsListRecyclerAdapter(private val contactsList: MutableList<ContactsModel> = ArrayList())
    : RecyclerView.Adapter<ContactsListRecyclerAdapter.ContactsHolder>(),
        ItemTouchHelperAdapter {

    var onItemClick: ((contactId: UUID) -> Unit)? = null
    var onItemSwipe: ((contactId: UUID) -> Unit)? = null

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ContactsListRecyclerAdapter.ContactsHolder {
        val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.item_contacts_list, viewGroup, false)
        return ContactsHolder(view, onItemClick!!)
    }

    override fun onBindViewHolder(contactsHolder: ContactsHolder, i: Int) {
        contactsHolder.bindContact(contactsList[i])
    }

    override fun getItemCount() = contactsList.size

    fun updateList(contactsList: List<ContactsModel>) {
        this.contactsList.clear()
        this.contactsList.addAll(contactsList)
        notifyDataSetChanged()
        Log.i(TAG, " notifyDataSetChanged() адаптера сработал")
    }

    override fun onItemDismiss(position: Int) {
        val contactId = contactsList[position].contactId
        onItemSwipe!!(contactId)
        contactsList.removeAt(position)
        notifyDataSetChanged()
    }


    inner class ContactsHolder(view: View, private val onItemClick: (contactId: UUID) -> Unit)
        : RecyclerView.ViewHolder(view){

        internal fun bindContact(contactsModel: ContactsModel) {
            with(contactsModel){
                itemView.setOnClickListener { onItemClick(contactId) }

                if (photoUri.isBlank()) {
                    itemView.contact_photo.setActualImageResource(R.drawable.ic_contact_photo)
                } else {
                    itemView.contact_photo.setImageURI(photoUri)
                }

                if (firstName.isBlank() && lastName.isBlank()) {
                    itemView.contact_name.setText(R.string.no_name)
                } else {
                    val fullName = "$lastName $firstName".trim()
                    itemView.contact_name.text = fullName
                }

                itemView.contact_have_telephone.visibility =
                if (contactsModel.telNumber.isBlank()) ImageView.GONE
                else ImageView.VISIBLE
            }
        }
    }
}